File: LeetCode Solutions/Recursion I/Swap Nodes in Pairs.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:
        if(head is None):
            return None
        if(head.next is None):
            return head
        if(head.next.next is None):
            newHeader = head.next
            newHeader.next = head
            head.next = None
            return newHeader
        nextPairHeader = head.next.next
        newHeader = head.next
        newHeader.next = head
        head.next = self.swapPairs(nextPairHeader)
        return newHeader

File: LeetCode Solutions/Recursion I/Swap Nodes in Pairs.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode swapPairs(ListNode head) {
        if(head == null){
            return null;
        }
        if(head.next == null){
            return head;
        }
        if(head.next.next == null){
            ListNode newHeader = head.next;
            newHeader.next = head;
            head.next = null;
            return newHeader;
        }
        ListNode nextPairHeader = head.next.next;
        ListNode newHeader = head.next;
        newHeader.next = head;
        head.next = swapPairs(nextPairHeader);
        return newHeader;
    }
}